lib/resty/etcd/v3.lua

```diff
@@ -688,43 +689,77 @@ local function request_chunk(self, method, path, opts, timeout)
 
 
     local function read_watch()
+        body = nil
 
         while(1) do
-            body, err = res.body_reader()
+            local chunk, error = res.body_reader()
+            if error then
+                return nil, error
+            end
+            if not chunk then
+                break
+            end
+
             if not body then
-                return nil, err
+                body = chunk
+            else
+                -- this branch will only be executed in rare cases, for example, a single event json
+                -- is larger than the proxy_buffer_size of nginx which proxies etcd, so it would be
+                -- ok to use a string concat directly without worry about the performance.
+                body = body .. chunk
             end
-            if not utils.is_empty_str(body) then
+
+            if not utils.is_empty_str(chunk) and str_byte(chunk, -1) == str_byte("\n") then
                 break
             end
 
         end
 
-        body, err = decode_json(body)
         if not body then
-            return nil, "failed to decode json body: " .. (err or " unkwon")
-        elseif body.error and body.error.http_code >= 500 then
-            -- health_check retry should do nothing here
-            -- and let connection closed to create a new one
-            health_check.report_failure(endpoint.http_host)
-            return nil, endpoint.http_host .. ": " .. body.error.http_status
+          return nil, nil
         end
 
-        if body.result and body.result.events then
-            for _, event in ipairs(body.result.events) do
-                if event.kv.value then   -- DELETE not have value
-                    event.kv.value = decode_base64(event.kv.value or "")
-                    event.kv.value = self.serializer.deserialize(event.kv.value)
+        local chunks, error = split(body, [[\n]], "jo")
```

Comment:
nice catch

Reply:
OK, but since `err` is already defined outside of `read_watch`, if I use `err`, our lint will complain about duplicate variable names, so I replace `error` with  `split_err`.